src/reducers-actions/orders.ts

```diff
@@ -126,3 +129,29 @@ export const reducer = (state: OrdersState, action: ReducerActionType): OrdersSt
       return state
   }
 }
+
+export async function sideEffect(state: OrdersState, action: ReducerActionType): Promise<void> {
+  switch (action.type) {
+    case 'OVERWRITE_ORDERS':
+    case 'UPDATE_ORDERS':
+    case 'APPEND_ORDERS':
+      const networkId = await walletApi.getNetworkId()
+      if (networkId) {
+        const newTokenIdsFromOrders = new Set<number>()
+
+        // orders can contain many duplicated tokenIds
+        state.orders.forEach(({ sellTokenId, buyTokenId }) => newTokenIdsFromOrders.add(sellTokenId).add(buyTokenId))
+
+        const tokensFromExchange = await Promise.all(
+          // calls are already cached, so we don't get duplicate calls later in components
+          Array.from(newTokenIdsFromOrders).map(tokenId => getTokenFromExchangeById({ tokenId, networkId })),
+        )
+
+        // only unlisted tokens get added inside
+        tokenListApi.addTokens({
+          tokens: tokensFromExchange.filter<TokenDetails>((token): token is TokenDetails => !!token),
+          networkId,
+        })
```

Comment:
This can be a service, if you want
@alfetopito @anxolin

Reply:
Whatever works to make this reusable, I'll be happy :)